File: analyses/methylation-summary/04-beta-tpm-correlation.R
# Calculate representative probe-level correlations between RNA-Seq (TPM) and 
# methylation (Beta) for patients who have samples in both datasets 

# Eric Wafula for Pediatric OpenTargets
# 03/18/2022

# Load libraries
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(tidyverse))

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# establish base dir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

# Set path to module and results directories
data_dir <- file.path(root_dir, "data")
module_dir <- file.path(root_dir, "analyses", "methylation-summary")
results_dir <- file.path(module_dir, "results")

# Get required columns from histologies file for RNA-Seq and methyalation 
# sample IDs
required_cols <- c("Kids_First_Biospecimen_ID", "Kids_First_Participant_ID", 
                   "experimental_strategy", "sample_type", "tumor_descriptor",
                   "cohort", "cancer_group")
histologies <- data.table::fread(file.path(data_dir, "histologies.tsv"), 
                                 sep = "\t", 
                                 select = required_cols, 
                                 showProgress = FALSE) %>% 
  tibble::as_tibble() %>% 
  dplyr::filter(sample_type == "Tumor" & 
                  (tumor_descriptor == "Primary Tumor" | 
                     tumor_descriptor == "Initial CNS Tumor") & 
                  (experimental_strategy == "RNA-Seq" | 
                     experimental_strategy == "Methylation"))

# Get methylation beta values
beta <- readr::read_rds(file.path(results_dir, "methyl-beta-values.rds"))

# Get RNA-Seq expression TPM values for array probes in GENCODE version 38 
# (Ensembl 104) gene symbols
tpm <- readr::read_rds(file.path(data_dir, 
                                 "gene-expression-rsem-tpm-collapsed.rds")) %>% 
  tibble::rownames_to_column(var = "Gene_symbol")
tpm <- data.table::fread(
  file.path(results_dir, "methyl-probe-annotations.tsv.gz"), 
  select = c("Probe_ID", "Gene_symbol"), showProgress = FALSE)  %>% 
  tibble::as_tibble() %>% 
  dplyr::distinct()  %>%
  dplyr::inner_join(tpm, by = "Gene_symbol") %>% 
  dplyr::select(-Gene_symbol)

# Calculating representative probe-level correlations between RNA-Seq (TPM) and 
# preprocessed methylation (Beta) samples
message("========================================================================================")
message("Calculating probe-level correlations between methylation beta and expression tpm values")
message("========================================================================================\n")
# Compute cancer type methylation to expression correlation
rnaseq_histologies <- histologies %>% 
  dplyr::filter(experimental_strategy == "RNA-Seq")
meth_histologies <- histologies %>% 
  dplyr::filter(experimental_strategy == "Methylation")
rm(histologies)
beta_tmp_correlation <- tibble::tibble()
for (cancer_type in unique(meth_histologies$cancer_group)) {
  message(c("Calculating probe-level correlations between beta and tpm values, ", cancer_type, " samples...\n"))
  # Get cancer type patients with RNA-Seq data
  cancer_type_rnaseq_ids <-  rnaseq_histologies %>% 
    dplyr::filter(cancer_group == cancer_type) %>% 
    dplyr::select(Kids_First_Biospecimen_ID, 
                  Kids_First_Participant_ID, cohort) %>% 
    dplyr::rename(RNASeq_ID = Kids_First_Biospecimen_ID)
  cancer_type_ids <- meth_histologies %>% 
    dplyr::filter(cancer_group == cancer_type) %>% 
    dplyr::select(Kids_First_Biospecimen_ID, 
                  Kids_First_Participant_ID,  cohort) %>% 
    dplyr::rename(Meth_ID = Kids_First_Biospecimen_ID) %>% 
    dplyr::inner_join(cancer_type_rnaseq_ids, 
                      by = c("Kids_First_Participant_ID", "cohort")) %>% 
    dplyr::rename(Patient_ID = Kids_First_Participant_ID) %>%
    dplyr::select(-cohort)
 
  # Get cancer type tpm values for patients with both RNA-Seq and 
  # methylation data and only keep array probes present in cancer type
  #  that have standard deviation == 0
  cancer_type_tpm <- tpm %>% 
    dplyr::select(tidyselect::any_of(c("Probe_ID",
                                       unique(cancer_type_ids$RNASeq_ID)))) %>%
    tidyr::pivot_longer(-Probe_ID, names_to = "RNASeq_ID", values_to = "TPM") %>% 
    dplyr::left_join(cancer_type_ids, by = "RNASeq_ID") %>% 
    dplyr::select(-c("Meth_ID", "RNASeq_ID")) %>%
    tidyr::pivot_wider(names_from = Patient_ID, 
                       values_from = TPM, values_fn = median) %>% 
    tidyr::pivot_longer(-Probe_ID, 
                        names_to = "Patient_ID", values_to = "TPM") %>%
    tidyr::pivot_wider(names_from = Probe_ID, values_from = TPM) %>% 
    purrr::discard(~ all(is.numeric(.x) & sd(.x) == 0))

  # Get cancer type beta values for patients with both RNA-Seq and 
  # methylation data and only keep array probes present in cancer type
  # tmp values that have standard deviation > 0
  cancer_type_beta <- beta %>% 
    dplyr::select(tidyselect::any_of(c("Probe_ID", 
                                       unique(cancer_type_ids$Meth_ID)))) %>%
    dplyr::filter(!is.na(.)) %>% 
    dplyr::filter(Probe_ID %in% unique(colnames(cancer_type_tpm))) %>% 
    tidyr::pivot_longer(-Probe_ID, names_to = "Meth_ID", values_to = "Beta") %>%
    dplyr::left_join(cancer_type_ids, by = "Meth_ID") %>% 
    dplyr::select(-c("Meth_ID", "RNASeq_ID")) %>%
    tidyr::pivot_wider(names_from = Patient_ID, 
                       values_from = Beta, values_fn = median) %>% 
    tidyr::pivot_longer(-Probe_ID, 
                        names_to = "Patient_ID", values_to = "Beta") %>%
    tidyr::pivot_wider(names_from = Probe_ID, values_from = Beta) %>% 
    purrr::discard(~ all(is.numeric(.x) & sd(.x) == 0))

  # Only keep cancer type tpm values for patient IDs and array probes that are 
  # in beta values then sort patient IDs and array probes to be in the same
  # order in both matrices. 
  # Once again drop any array probes with standard deviation == 0 as result of
  # filtering 
  cancer_type_tpm <- cancer_type_tpm %>% 
    dplyr::filter(Patient_ID %in% cancer_type_beta$Patient_ID) %>%
    purrr::discard(~ all(is.numeric(.x) & sd(.x) == 0)) %>% 
    dplyr::arrange(Patient_ID)
  cancer_type_beta <- cancer_type_beta %>% 
    dplyr::select(tidyselect::any_of(colnames(cancer_type_tpm))) %>%
    dplyr::arrange(Patient_ID) %>%
    tibble::column_to_rownames(var = "Patient_ID") %>% 
    dplyr::select(sort(tidyselect::peek_vars()))
  cancer_type_tpm <- cancer_type_tpm %>% 
    tibble::column_to_rownames(var = "Patient_ID") %>%
    dplyr::select(tidyselect::all_of(colnames(cancer_type_beta))) %>%
    dplyr::select(sort(tidyselect::peek_vars()))
  
  # calculate probe correlation between methylation beta values
  # RNA-Seq expression tpm values
  beta_tmp_correlation <-  beta_tmp_correlation %>%
    dplyr::bind_rows(
      tibble::tibble(
        "Probe_ID" = names(cancer_type_beta),
        "RNA_Correlation" = 
          sapply(1:ncol(cancer_type_beta),
                 function(i) cor(cancer_type_beta[,i], cancer_type_tpm[,i])),
      ) %>% 
        dplyr::mutate(Disease = cancer_type)
    )
  rm(cancer_type_rnaseq_ids, cancer_type_ids, cancer_type_beta, cancer_type_tpm)
}
rm(beta, tpm, rnaseq_histologies, meth_histologies) 

# Write probe-level correlations between methylation beta values
# RNA-Seq expression tpm values
message("Writing probe-level correlations to methyl-probe-beta-tpm-correlations.tsv file...\n")
beta_tmp_correlation %>% data.table::setDT() %>%
  data.table::fwrite(file.path(results_dir,
                               "methyl-probe-beta-tpm-correlations.tsv.gz"), 
                     sep="\t", compress = "auto")

message("Analysis Done..\n")

File: analyses/methylation-summary/04-beta-tpm-correlation.py
#!/usr/bin/env python3


"""
04-beta-tpm-correlation.py
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Calculate representative probe-level correlations between RNA-Seq (TPM) and methylation (Beta) for patients who have samples in both datasets
"""


__author__ = ('Eric Wafula (wafulae@chop.edu)')
__version__ = '1.0'
__date__ = '15 May 2022'


import re
import os
import sys
import git
import numpy as np
import pandas as pd
import pyreadr as rds


# establish base dir
root_dir = git.Repo('.', search_parent_directories=True).working_tree_dir

# Set path to module and results directories
data_dir = os.path.join(root_dir, "data")
module_dir = os.path.join(root_dir, "analyses", "methylation-summary")
results_dir = os.path.join(module_dir, "results")

print(f"\nStarting Analysis...\n")

# Get required columns from histologies file for RNA-Seq and methyalation sample IDs
hist_cols = ["Kids_First_Biospecimen_ID", "Kids_First_Participant_ID", "experimental_strategy", "sample_type", "tumor_descriptor","cohort", "cancer_group"]
histologies = pd.read_csv(os.path.join(data_dir, "histologies.tsv"), usecols = hist_cols, sep="\t", na_filter=False, dtype=str)
rnaseq_histologies = histologies.loc[((histologies.tumor_descriptor == "Primary Tumor") | (histologies.tumor_descriptor == "Initial CNS Tumor")) & (histologies.experimental_strategy == "RNA-Seq")]
meth_histologies = histologies.loc[((histologies.tumor_descriptor == "Primary Tumor") | (histologies.tumor_descriptor == "Initial CNS Tumor")) & (histologies.experimental_strategy == "Methylation")]
del(histologies)

# Get methylation beta values
beta = rds.read_r(os.path.join(results_dir, "methyl-beta-values.rds"))[None]

# Get RNA-Seq expression TPM values for array probes in GENCODE version 38 (Ensembl 104) gene symbols
tpm = rds.read_r(os.path.join(data_dir,"gene-expression-rsem-tpm-collapsed.rds"))[None].reset_index()
tpm.rename(columns={"rownames": "Gene_symbol"}, inplace = True)

# Get array probes in GENCODE version 38 (Ensembl 104) gene symbols
annot_cols = ["Probe_ID", "Gene_symbol"]
probe_annot = pd.read_csv(os.path.join(results_dir, "methyl-probe-annotations.tsv.gz"), usecols = annot_cols, sep="\t", na_filter=False, dtype=str).drop_duplicates().reset_index(drop=True)

# merge probe annotations to tpm matrix
tpm = pd.merge(probe_annot, tpm, how = "inner", on = "Gene_symbol")
tpm = tpm.loc[:, tpm.columns != "Gene_symbol"]

# Calculating representative probe-level correlations between RNA-Seq (TPM) and preprocessed methylation (Beta) samples
print(f"========================================================================================")
print(f"Calculating probe-level correlations between methylation beta and expression tpm values")
print(f"========================================================================================\n")
merging_list = []
for cohort in meth_histologies.cohort.unique():
	cohort_type_meth_ids = meth_histologies.loc[meth_histologies.cohort == cohort]
	for cancer_type in cohort_type_meth_ids.cancer_group.unique():
		print(f"Calculating probe-level correlations between beta and tpm values for {cancer_type} cancer group samples in {cohort} cohort...\n")
		# Get cancer type patients with both methylation and RNA-Seq data
		cancer_type_rnaseq_ids = rnaseq_histologies.loc[(rnaseq_histologies.cohort == cohort) & (rnaseq_histologies.cancer_group == cancer_type)]
		cancer_type_rnaseq_ids = cancer_type_rnaseq_ids[["Kids_First_Biospecimen_ID", "Kids_First_Participant_ID"]]
		cancer_type_rnaseq_ids.rename(columns={"Kids_First_Biospecimen_ID": "RNASeq_ID"}, inplace = True)
		cancer_type_meth_ids = meth_histologies.loc[(meth_histologies.cohort == cohort) & (meth_histologies.cancer_group == cancer_type)]
		cancer_type_meth_ids = cancer_type_meth_ids[["Kids_First_Biospecimen_ID", "Kids_First_Participant_ID"]]
		cancer_type_meth_ids.rename(columns={"Kids_First_Biospecimen_ID": "Meth_ID"}, inplace = True)
		cancer_type_ids = pd.merge(cancer_type_meth_ids, cancer_type_rnaseq_ids, how = "inner", on = "Kids_First_Participant_ID")
		cancer_type_ids.rename(columns={"Kids_First_Participant_ID": "Patient_ID"}, inplace = True)
		if cancer_type_ids.empty:
			continue

		# Get cancer type beta values for patients with both RNA-Seq and methylation data
		sample_ids = np.append(["Probe_ID"], cancer_type_ids.Meth_ID.unique()).tolist()
		cancer_type_beta = beta.loc[:,beta.columns.isin(sample_ids)].groupby("Probe_ID", as_index=False).median()
		cancer_type_beta = cancer_type_beta.set_index("Probe_ID").rename_axis(None)
		cancer_type_beta = cancer_type_beta.transpose().reset_index()
		cancer_type_beta.rename(columns={"index": "Meth_ID"}, inplace = True)
		cancer_type_beta.dropna(axis = 1, how = "all", inplace = True)
		cancer_type_beta = pd.merge(cancer_type_ids, cancer_type_beta, how = "inner", on = "Meth_ID")
		cancer_type_beta = cancer_type_beta.loc[:, ~cancer_type_beta.columns.isin(["Meth_ID", "RNASeq_ID"])]
		cancer_type_beta = cancer_type_beta.groupby("Patient_ID", as_index=False).median()

		# Get cancer type tpm values for patients with both RNA-Seq and methylation data
		sample_ids = np.append(["Probe_ID"], cancer_type_ids.RNASeq_ID.unique()).tolist()
		cancer_type_tpm = tpm.loc[:,tpm.columns.isin(sample_ids)].groupby("Probe_ID", as_index=False).median()
		cancer_type_tpm = cancer_type_tpm.set_index("Probe_ID").rename_axis(None)
		cancer_type_tpm = cancer_type_tpm.transpose().reset_index()
		cancer_type_tpm.rename(columns={"index": "RNASeq_ID"}, inplace = True)
		cancer_type_tpm.dropna(axis = 1, how = "all", inplace = True)
		cancer_type_tpm = pd.merge(cancer_type_ids, cancer_type_tpm, how = "inner", on = "RNASeq_ID")
		cancer_type_tpm = cancer_type_tpm.loc[:, ~cancer_type_tpm.columns.isin(["Meth_ID", "RNASeq_ID"])]
		cancer_type_tpm = cancer_type_tpm.groupby("Patient_ID", as_index=False).median()

		# calculate probe correlation between methylation beta values RNA-Seq expression tpm values
		cancer_type_beta = cancer_type_beta.sort_values(by = "Patient_ID").set_index("Patient_ID").rename_axis(None)
		cancer_type_tpm = cancer_type_tpm.sort_values(by = "Patient_ID").set_index("Patient_ID").rename_axis(None)
		cancer_type_correlation = cancer_type_beta.corrwith(cancer_type_tpm, drop = True).dropna().to_frame().reset_index()
		cancer_type_correlation.rename(columns = {0: "RNA_Correlation", "index": "Probe_ID"}, inplace = True)
		cancer_type_correlation["Dataset"] = cohort
		cancer_type_correlation["Disease"] = cancer_type
		if not cancer_type_correlation.empty:
			merging_list.append(cancer_type_correlation)

# Write probe-level correlations between methylation beta values RNA-Seq expression tpm values
print(f"Writing probe-level correlations to methyl-probe-beta-tpm-correlations.tsv file...\n")
beta_tmp_correlation = pd.concat(merging_list, sort=False, ignore_index=True)
beta_tmp_correlation.to_csv(os.path.join(results_dir,"methyl-probe-beta-tpm-correlations.tsv.gz"), sep="\t", index=False, encoding="utf-8")

print(f"Analysis Done...\n")

